File: skills/testing/unit-testing/templates/example.test.js
/**
 * Unit Testing Examples - JavaScript/Jest
 *
 * This file demonstrates best practices for writing unit tests with Jest,
 * including mocking, async testing, snapshot testing, and test coverage.
 *
 * @see https://jestjs.io/docs/getting-started
 */

// ===== Example Code Under Test =====

class Calculator {
  add(a, b) {
    return a + b;
  }

  subtract(a, b) {
    return a - b;
  }

  divide(a, b) {
    if (b === 0) {
      throw new Error('Cannot divide by zero');
    }
    return a / b;
  }

  multiply(a, b) {
    return a * b;
  }
}

class User {
  constructor(id, name, email, isActive = true) {
    this.id = id;
    this.name = name;
    this.email = email;
    this.isActive = isActive;
  }

  deactivate() {
    this.isActive = false;
  }

  getDisplayName() {
    return `${this.name} (${this.email})`;
  }
}

class UserRepository {
  constructor(database) {
    this.db = database;
  }

  async getUserById(userId) {
    const result = await this.db.query(`SELECT * FROM users WHERE id = ${userId}`);
    return new User(result.id, result.name, result.email, result.is_active);
  }

  async saveUser(user) {
    return await this.db.execute(`INSERT INTO users VALUES ${user.id}`);
  }
}

class ApiService {
  async fetchData(url) {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return await response.json();
  }
}

// ===== Basic Unit Tests =====

describe('Calculator', () => {
  let calculator;

  // Setup before each test
  beforeEach(() => {
    calculator = new Calculator();
  });

  // Teardown after each test
  afterEach(() => {
    // Clean up if needed
  });

  describe('add', () => {
    test('should add two positive numbers', () => {
      const result = calculator.add(2, 3);
      expect(result).toBe(5);
    });

    test('should add negative numbers', () => {
      const result = calculator.add(-2, -3);
      expect(result).toBe(-5);
    });

    test('should add zero', () => {
      const result = calculator.add(5, 0);
      expect(result).toBe(5);
    });
  });

  describe('subtract', () => {
    test('should subtract two numbers', () => {
      const result = calculator.subtract(10, 3);
      expect(result).toBe(7);
    });
  });

  describe('multiply', () => {
    test('should multiply two numbers', () => {
      const result = calculator.multiply(4, 5);
      expect(result).toBe(20);
    });
  });

  describe('divide', () => {
    test('should divide two numbers', () => {
      const result = calculator.divide(10, 2);
      expect(result).toBe(5);
    });

    test('should throw error when dividing by zero', () => {
      expect(() => calculator.divide(10, 0)).toThrow('Cannot divide by zero');
    });
  });
});

// ===== Parametrized Tests (using test.each) =====

describe('Calculator - Parametrized Tests', () => {
  const calculator = new Calculator();

  test.each([
    [2, 3, 5],
    [0, 0, 0],
    [-1, 1, 0],
    [100, 200, 300],
    [-5, -3, -8],
  ])('add(%i, %i) should return %i', (a, b, expected) => {
    expect(calculator.add(a, b)).toBe(expected);
  });

  test.each([
    [10, 2, 5],
    [100, 10, 10],
    [7, 2, 3.5],
    [-10, 2, -5],
  ])('divide(%i, %i) should return %i', (a, b, expected) => {
    expect(calculator.divide(a, b)).toBe(expected);
  });

  test.each([0, -0])('divide(10, %i) should throw error', (divisor) => {
    expect(() => calculator.divide(10, divisor)).toThrow();
  });
});

// ===== Mocking Tests =====

describe('UserRepository', () => {
  let mockDatabase;
  let userRepository;

  beforeEach(() => {
    // Create mock database
    mockDatabase = {
      query: jest.fn(),
      execute: jest.fn(),
    };

    userRepository = new UserRepository(mockDatabase);
  });

  describe('getUserById', () => {
    test('should retrieve user from database', async () => {
      // Arrange
      const mockUser = {
        id: 1,
        name: 'Alice',
        email: 'alice@example.com',
        is_active: true,
      };
      mockDatabase.query.mockResolvedValue(mockUser);

      // Act
      const user = await userRepository.getUserById(1);

      // Assert
      expect(mockDatabase.query).toHaveBeenCalledWith('SELECT * FROM users WHERE id = 1');
      expect(mockDatabase.query).toHaveBeenCalledTimes(1);
      expect(user).toBeInstanceOf(User);
      expect(user.id).toBe(1);
      expect(user.name).toBe('Alice');
      expect(user.email).toBe('alice@example.com');
      expect(user.isActive).toBe(true);
    });

    test('should handle database errors', async () => {
      // Arrange
      mockDatabase.query.mockRejectedValue(new Error('Database connection failed'));

      // Act & Assert
      await expect(userRepository.getUserById(1)).rejects.toThrow(
        'Database connection failed'
      );
    });
  });

  describe('saveUser', () => {
    test('should save user to database', async () => {
      // Arrange
      const user = new User(1, 'Bob', 'bob@example.com');
      mockDatabase.execute.mockResolvedValue(true);

      // Act
      const result = await userRepository.saveUser(user);

      // Assert
      expect(mockDatabase.execute).toHaveBeenCalledTimes(1);
      expect(result).toBe(true);
    });
  });
});

// ===== Async Tests =====

describe('ApiService', () => {
  let apiService;

  beforeEach(() => {
    apiService = new ApiService();
  });

  describe('fetchData', () => {
    test('should fetch data successfully', async () => {
      // Arrange
      const mockData = { id: 1, name: 'Test' };
      global.fetch = jest.fn(() =>
        Promise.resolve({
          ok: true,
          json: () => Promise.resolve(mockData),
        })
      );

      // Act
      const result = await apiService.fetchData('https://api.example.com/data');

      // Assert
      expect(fetch).toHaveBeenCalledWith('https://api.example.com/data');
      expect(result).toEqual(mockData);
    });

    test('should throw error on HTTP failure', async () => {
      // Arrange
      global.fetch = jest.fn(() =>
        Promise.resolve({
          ok: false,
          status: 404,
        })
      );

      // Act & Assert
      await expect(apiService.fetchData('https://api.example.com/data')).rejects.toThrow(
        'HTTP error! status: 404'
      );
    });

    test('should handle network errors', async () => {
      // Arrange
      global.fetch = jest.fn(() => Promise.reject(new Error('Network error')));

      // Act & Assert
      await expect(apiService.fetchData('https://api.example.com/data')).rejects.toThrow(
        'Network error'
      );
    });
  });
});

// ===== User Class Tests =====

describe('User', () => {
  describe('constructor', () => {
    test('should create user with correct properties', () => {
      const user = new User(1, 'Alice', 'alice@example.com');

      expect(user.id).toBe(1);
      expect(user.name).toBe('Alice');
      expect(user.email).toBe('alice@example.com');
      expect(user.isActive).toBe(true);
    });

    test('should create inactive user when specified', () => {
      const user = new User(1, 'Bob', 'bob@example.com', false);

      expect(user.isActive).toBe(false);
    });
  });

  describe('deactivate', () => {
    test('should deactivate user', () => {
      const user = new User(1, 'Alice', 'alice@example.com');

      user.deactivate();

      expect(user.isActive).toBe(false);
    });
  });

  describe('getDisplayName', () => {
    test('should return formatted display name', () => {
      const user = new User(1, 'Alice', 'alice@example.com');

      const displayName = user.getDisplayName();

      expect(displayName).toBe('Alice (alice@example.com)');
    });

    test('should match snapshot', () => {
      const user = new User(1, 'Alice', 'alice@example.com');

      expect(user.getDisplayName()).toMatchSnapshot();
    });
  });
});

// ===== Spy Tests =====

describe('Spies', () => {
  test('should spy on method calls', () => {
    const user = new User(1, 'Alice', 'alice@example.com');
    const spy = jest.spyOn(user, 'getDisplayName');

    user.getDisplayName();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenCalledTimes(1);

    spy.mockRestore(); // Clean up spy
  });

  test('should mock method return value', () => {
    const user = new User(1, 'Alice', 'alice@example.com');
    const spy = jest.spyOn(user, 'getDisplayName').mockReturnValue('Mocked Name');

    const displayName = user.getDisplayName();

    expect(displayName).toBe('Mocked Name');

    spy.mockRestore();
  });
});

// ===== Timer Mocks =====

describe('Timer Mocks', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('should execute callback after timeout', () => {
    const callback = jest.fn();

    setTimeout(callback, 1000);

    expect(callback).not.toHaveBeenCalled();

    jest.advanceTimersByTime(1000);

    expect(callback).toHaveBeenCalled();
    expect(callback).toHaveBeenCalledTimes(1);
  });

  test('should execute multiple timeouts', () => {
    const callback1 = jest.fn();
    const callback2 = jest.fn();

    setTimeout(callback1, 1000);
    setTimeout(callback2, 2000);

    jest.advanceTimersByTime(1500);

    expect(callback1).toHaveBeenCalled();
    expect(callback2).not.toHaveBeenCalled();

    jest.advanceTimersByTime(500);

    expect(callback2).toHaveBeenCalled();
  });
});

// ===== Test Organization Best Practices =====
//
// 1. Use describe blocks to group related tests
// 2. Use beforeEach/afterEach for setup/teardown
// 3. Use clear, descriptive test names
// 4. Follow AAA pattern: Arrange, Act, Assert
// 5. Keep tests independent and isolated
// 6. Mock external dependencies
// 7. Test edge cases and error conditions
// 8. Use parametrized tests for similar test cases
// 9. Aim for >80% code coverage
// 10. Keep tests fast and focused

// ===== Jest Matchers Reference =====
//
// Common Matchers:
// - expect(value).toBe(expected)              // Strict equality (===)
// - expect(value).toEqual(expected)           // Deep equality
// - expect(value).toBeTruthy()                // Truthy value
// - expect(value).toBeFalsy()                 // Falsy value
// - expect(value).toBeNull()                  // Null value
// - expect(value).toBeUndefined()             // Undefined value
// - expect(value).toBeDefined()               // Defined value
// - expect(fn).toThrow()                      // Function throws error
// - expect(fn).toThrow('error message')       // Function throws specific error
// - expect(array).toContain(item)             // Array contains item
// - expect(string).toMatch(/regex/)           // String matches regex
// - expect(fn).toHaveBeenCalled()             // Mock was called
// - expect(fn).toHaveBeenCalledWith(arg)      // Mock called with argument
// - expect(fn).toHaveBeenCalledTimes(n)       // Mock called n times
// - expect(promise).resolves.toBe(value)      // Promise resolves to value
// - expect(promise).rejects.toThrow()         // Promise rejects with error

module.exports = {
  Calculator,
  User,
  UserRepository,
  ApiService,
};

File: skills/testing/unit-testing/templates/example_test.go
// Unit Testing Examples - Go
//
// This file demonstrates best practices for writing unit tests in Go,
// including table-driven tests, mocking, benchmarking, and test coverage.
//
// See: https://pkg.go.dev/testing

package testing_examples

import (
	"errors"
	"testing"
)

// ===== Example Code Under Test =====

// Calculator provides basic arithmetic operations
type Calculator struct{}

func (c *Calculator) Add(a, b int) int {
	return a + b
}

func (c *Calculator) Subtract(a, b int) int {
	return a - b
}

func (c *Calculator) Divide(a, b int) (float64, error) {
	if b == 0 {
		return 0, errors.New("cannot divide by zero")
	}
	return float64(a) / float64(b), nil
}

func (c *Calculator) Multiply(a, b int) int {
	return a * b
}

// User represents a user in the system
type User struct {
	ID       int
	Name     string
	Email    string
	IsActive bool
}

func (u *User) Deactivate() {
	u.IsActive = false
}

func (u *User) GetDisplayName() string {
	return u.Name + " (" + u.Email + ")"
}

// Database interface for mocking
type Database interface {
	Query(query string) (map[string]interface{}, error)
	Execute(query string) error
}

// UserRepository handles user data access
type UserRepository struct {
	db Database
}

func NewUserRepository(db Database) *UserRepository {
	return &UserRepository{db: db}
}

func (r *UserRepository) GetUserByID(id int) (*User, error) {
	result, err := r.db.Query("SELECT * FROM users WHERE id = ?")
	if err != nil {
		return nil, err
	}

	user := &User{
		ID:       result["id"].(int),
		Name:     result["name"].(string),
		Email:    result["email"].(string),
		IsActive: result["is_active"].(bool),
	}

	return user, nil
}

func (r *UserRepository) SaveUser(user *User) error {
	return r.db.Execute("INSERT INTO users VALUES (?)")
}

// ===== Basic Unit Tests =====

func TestCalculatorAdd(t *testing.T) {
	calc := &Calculator{}
	result := calc.Add(2, 3)

	if result != 5 {
		t.Errorf("Add(2, 3) = %d; want 5", result)
	}
}

func TestCalculatorSubtract(t *testing.T) {
	calc := &Calculator{}
	result := calc.Subtract(10, 3)

	if result != 7 {
		t.Errorf("Subtract(10, 3) = %d; want 7", result)
	}
}

func TestCalculatorMultiply(t *testing.T) {
	calc := &Calculator{}
	result := calc.Multiply(4, 5)

	if result != 20 {
		t.Errorf("Multiply(4, 5) = %d; want 20", result)
	}
}

func TestCalculatorDivide(t *testing.T) {
	calc := &Calculator{}
	result, err := calc.Divide(10, 2)

	if err != nil {
		t.Errorf("Divide(10, 2) returned error: %v", err)
	}

	if result != 5.0 {
		t.Errorf("Divide(10, 2) = %f; want 5.0", result)
	}
}

func TestCalculatorDivideByZero(t *testing.T) {
	calc := &Calculator{}
	_, err := calc.Divide(10, 0)

	if err == nil {
		t.Error("Divide(10, 0) should return an error")
	}

	expectedError := "cannot divide by zero"
	if err.Error() != expectedError {
		t.Errorf("Divide(10, 0) error = %q; want %q", err.Error(), expectedError)
	}
}

// ===== Table-Driven Tests (Go Best Practice) =====

func TestCalculatorAddTableDriven(t *testing.T) {
	calc := &Calculator{}

	tests := []struct {
		name     string
		a        int
		b        int
		expected int
	}{
		{"positive numbers", 2, 3, 5},
		{"zeros", 0, 0, 0},
		{"negative and positive", -1, 1, 0},
		{"large numbers", 100, 200, 300},
		{"negative numbers", -5, -3, -8},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := calc.Add(tt.a, tt.b)
			if result != tt.expected {
				t.Errorf("Add(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
			}
		})
	}
}

func TestCalculatorDivideTableDriven(t *testing.T) {
	calc := &Calculator{}

	tests := []struct {
		name        string
		a           int
		b           int
		expected    float64
		expectError bool
	}{
		{"normal division", 10, 2, 5.0, false},
		{"large numbers", 100, 10, 10.0, false},
		{"decimal result", 7, 2, 3.5, false},
		{"negative dividend", -10, 2, -5.0, false},
		{"divide by zero", 10, 0, 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := calc.Divide(tt.a, tt.b)

			if tt.expectError {
				if err == nil {
					t.Errorf("Divide(%d, %d) should return an error", tt.a, tt.b)
				}
			} else {
				if err != nil {
					t.Errorf("Divide(%d, %d) returned unexpected error: %v", tt.a, tt.b, err)
				}
				if result != tt.expected {
					t.Errorf("Divide(%d, %d) = %f; want %f", tt.a, tt.b, result, tt.expected)
				}
			}
		})
	}
}

// ===== Mock Database for Testing =====

type MockDatabase struct {
	QueryFunc   func(query string) (map[string]interface{}, error)
	ExecuteFunc func(query string) error
}

func (m *MockDatabase) Query(query string) (map[string]interface{}, error) {
	if m.QueryFunc != nil {
		return m.QueryFunc(query)
	}
	return nil, nil
}

func (m *MockDatabase) Execute(query string) error {
	if m.ExecuteFunc != nil {
		return m.ExecuteFunc(query)
	}
	return nil
}

// ===== Tests with Mocks =====

func TestUserRepositoryGetUserByID(t *testing.T) {
	mockDB := &MockDatabase{
		QueryFunc: func(query string) (map[string]interface{}, error) {
			return map[string]interface{}{
				"id":        1,
				"name":      "Alice",
				"email":     "alice@example.com",
				"is_active": true,
			}, nil
		},
	}

	repo := NewUserRepository(mockDB)
	user, err := repo.GetUserByID(1)

	if err != nil {
		t.Errorf("GetUserByID(1) returned error: %v", err)
	}

	if user.ID != 1 {
		t.Errorf("user.ID = %d; want 1", user.ID)
	}

	if user.Name != "Alice" {
		t.Errorf("user.Name = %q; want %q", user.Name, "Alice")
	}

	if user.Email != "alice@example.com" {
		t.Errorf("user.Email = %q; want %q", user.Email, "alice@example.com")
	}

	if !user.IsActive {
		t.Error("user.IsActive = false; want true")
	}
}

func TestUserRepositoryGetUserByIDError(t *testing.T) {
	mockDB := &MockDatabase{
		QueryFunc: func(query string) (map[string]interface{}, error) {
			return nil, errors.New("database connection failed")
		},
	}

	repo := NewUserRepository(mockDB)
	_, err := repo.GetUserByID(1)

	if err == nil {
		t.Error("GetUserByID(1) should return an error when database fails")
	}

	expectedError := "database connection failed"
	if err.Error() != expectedError {
		t.Errorf("error = %q; want %q", err.Error(), expectedError)
	}
}

func TestUserRepositorySaveUser(t *testing.T) {
	executeCalled := false

	mockDB := &MockDatabase{
		ExecuteFunc: func(query string) error {
			executeCalled = true
			return nil
		},
	}

	repo := NewUserRepository(mockDB)
	user := &User{ID: 1, Name: "Bob", Email: "bob@example.com", IsActive: true}

	err := repo.SaveUser(user)

	if err != nil {
		t.Errorf("SaveUser() returned error: %v", err)
	}

	if !executeCalled {
		t.Error("SaveUser() did not call database Execute()")
	}
}

// ===== User Tests =====

func TestUserDeactivate(t *testing.T) {
	user := &User{ID: 1, Name: "Alice", Email: "alice@example.com", IsActive: true}

	user.Deactivate()

	if user.IsActive {
		t.Error("user.IsActive = true; want false after Deactivate()")
	}
}

func TestUserGetDisplayName(t *testing.T) {
	user := &User{ID: 1, Name: "Alice", Email: "alice@example.com", IsActive: true}

	displayName := user.GetDisplayName()
	expected := "Alice (alice@example.com)"

	if displayName != expected {
		t.Errorf("GetDisplayName() = %q; want %q", displayName, expected)
	}
}

func TestUserGetDisplayNameTableDriven(t *testing.T) {
	tests := []struct {
		name     string
		user     User
		expected string
	}{
		{
			"normal user",
			User{ID: 1, Name: "Alice", Email: "alice@example.com", IsActive: true},
			"Alice (alice@example.com)",
		},
		{
			"user with special characters",
			User{ID: 2, Name: "Bob O'Brien", Email: "bob+test@example.com", IsActive: true},
			"Bob O'Brien (bob+test@example.com)",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := tt.user.GetDisplayName()
			if result != tt.expected {
				t.Errorf("GetDisplayName() = %q; want %q", result, tt.expected)
			}
		})
	}
}

// ===== Subtests =====

func TestUser(t *testing.T) {
	t.Run("Deactivate", func(t *testing.T) {
		user := &User{ID: 1, Name: "Alice", Email: "alice@example.com", IsActive: true}
		user.Deactivate()

		if user.IsActive {
			t.Error("user.IsActive = true; want false")
		}
	})

	t.Run("GetDisplayName", func(t *testing.T) {
		user := &User{ID: 1, Name: "Alice", Email: "alice@example.com", IsActive: true}
		displayName := user.GetDisplayName()

		if displayName != "Alice (alice@example.com)" {
			t.Errorf("GetDisplayName() = %q; want %q", displayName, "Alice (alice@example.com)")
		}
	})
}

// ===== Benchmarks =====

func BenchmarkCalculatorAdd(b *testing.B) {
	calc := &Calculator{}

	for i := 0; i < b.N; i++ {
		calc.Add(2, 3)
	}
}

func BenchmarkCalculatorDivide(b *testing.B) {
	calc := &Calculator{}

	for i := 0; i < b.N; i++ {
		calc.Divide(10, 2)
	}
}

func BenchmarkUserGetDisplayName(b *testing.B) {
	user := &User{ID: 1, Name: "Alice", Email: "alice@example.com", IsActive: true}

	for i := 0; i < b.N; i++ {
		user.GetDisplayName()
	}
}

// ===== Test Helpers =====

func assertEqual(t *testing.T, got, want interface{}) {
	t.Helper()
	if got != want {
		t.Errorf("got %v; want %v", got, want)
	}
}

func assertNoError(t *testing.T, err error) {
	t.Helper()
	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
}

func assertError(t *testing.T, err error, expectedMessage string) {
	t.Helper()
	if err == nil {
		t.Error("expected error, got nil")
		return
	}
	if err.Error() != expectedMessage {
		t.Errorf("error = %q; want %q", err.Error(), expectedMessage)
	}
}

// Example using test helpers
func TestCalculatorDivideWithHelpers(t *testing.T) {
	calc := &Calculator{}

	t.Run("normal division", func(t *testing.T) {
		result, err := calc.Divide(10, 2)
		assertNoError(t, err)
		assertEqual(t, result, 5.0)
	})

	t.Run("divide by zero", func(t *testing.T) {
		_, err := calc.Divide(10, 0)
		assertError(t, err, "cannot divide by zero")
	})
}

// ===== Testing Best Practices =====
//
// 1. Use table-driven tests for similar test cases
// 2. Use subtests (t.Run) to organize related tests
// 3. Use t.Helper() in test helper functions
// 4. Test both success and error paths
// 5. Use mocks/interfaces for dependencies
// 6. Write benchmarks for performance-critical code
// 7. Aim for high code coverage (use `go test -cover`)
// 8. Keep tests independent and isolated
// 9. Use descriptive test names
// 10. Run tests with race detector: `go test -race`
//
// Commands:
//   go test                          # Run all tests
//   go test -v                       # Verbose output
//   go test -cover                   # Show coverage
//   go test -coverprofile=cover.out  # Generate coverage report
//   go tool cover -html=cover.out    # View coverage in browser
//   go test -bench=.                 # Run benchmarks
//   go test -race                    # Run with race detector
//   go test -run TestName            # Run specific test
